### PR TITLE
Implement multiple span support for DocumentQuestionAnswering

### PR DESCRIPTION
**Pull request by ankrgyl**
# What does this PR do?

Extends DocumentQuestionAnswering pipeline to include support for multiple spans.

Fixes #18414

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@NielsRogge @sgugger @Narsil

### CODE REVIEW

**Comment by ankrgyl** [comment]
I still need to extend the tests to set the `max_seq_len` to be low (e.g. 10) and make sure that they work.

@NielsRogge another thing we could do in this change, or a separate one if y'all prefer, is to support multiple pages. The [DocQuery pipeline](https://github.com/impira/docquery/blob/main/src/docquery/ext/pipeline_document_question_answering.py#L227) already does this, because it's common to have multi-page documents and quite useful to be able to search across them. It's important to have this in the pipeline itself, so that you can properly handle the "no answer" score and top-k across pages. The main consideration is how we change the input shape of the pipeline to accommodate it, since we currently accept a (single) image and (optional) list of word_boxes. The options I can think of are:

- Image can also be a list of images, and then `word_boxes` can be a list of list of word boxes
- Combine image and `word_boxes` into one list of `(image, [(word, box)])` tuples
- Add a new optional argument, called pages, and use that instead of the image/word_box arguments if it's set.

File: src/transformers/pipelines/document_question_answering.py [hunk 1]
```diff
@@ -305,44 +329,49 @@ def preprocess(self, input, lang=None, tesseract_config=""):
...
 
-            # TODO: For now, this should always be num_spans == 1 given the flags we've passed in above, but the
-            # code is written to naturally handle multiple spans at the right time.
             num_spans = len(encoding["input_ids"])
 
             # p_mask: mask with 1 for token than cannot be in the answer (0 for token which can be in an answer)
             # We put 0 on the tokens from the context and 1 everywhere else (question and special tokens)
             # This logic mirrors the logic in the question_answering pipeline
             p_mask = [[tok != 1 for tok in encoding.sequence_ids(span_id)] for span_id in range(num_spans)]
+            print(num_spans)
```
Comment (sgugger) [inline src/transformers/pipelines/document_question_answering.py:350]: That's some debugging statement that should be removed, right?

File: src/transformers/pipelines/document_question_answering.py [hunk 2]
```diff
@@ -399,41 +433,38 @@ def postprocess_donut(self, model_outputs, **kwargs):
...
+            starts, ends, scores, min_null_score = select_starts_ends(
+                output["start_logits"],
+                output["end_logits"],
+                output["p_mask"],
+                output["attention_mask"].numpy() if output.get("attention_mask", None) is not None else None,
+                min_null_score,
+                top_k,
+                handle_impossible_answer,
+                max_answer_len,
+            )
```
Comment (sgugger) [inline src/transformers/pipelines/document_question_answering.py:455]: Nit, using keyword arguments here would make the code easier to maintain.

**Review by sgugger — APPROVED** [review]
I will let @NielsRogge dive deeper in this, but looking good on my side, thanks!

File: src/transformers/pipelines/document_question_answering.py (hunk 1 above)
Comment (ankrgyl) [reply src/transformers/pipelines/document_question_answering.py:350]: Oops will remove! I was manually verifying that it worked :)

File: src/transformers/pipelines/document_question_answering.py (hunk 2 above)
Comment (ankrgyl) [reply src/transformers/pipelines/document_question_answering.py:455]: I can change that! Good idea.

**Comment by ankrgyl** [comment]
Absolutely. Apologies for the delay -- I will update the PR within the next few days.

**Comment by ankrgyl** [comment]
@Narsil apologies again for the delay. I wanted to make sure I had some time to test it thoroughly. I added tests for the two flavors of LayoutLM models, but not for Donut, since there are no spans for Donut (i.e. there is no parameter or argument I can provide that hits two different code paths with the Chunk Pipeline, at least as far as I can tell).

I found several small bugs along the way and fixed them too.

**Review by Narsil — APPROVED** [review]
LGTM.

Thank you !

**Review by sgugger — APPROVED** [review]
Thnaks for working on this!

**Comment by ydshieh** [comment]
Hi @ankrgyl Thank you for this PR. We need some help for CI failure regarding this new addition 🙏 

After this PR, we have 5 CI failures
```bash
FAILED tests/pipelines/test_pipelines_document_question_answering.py::DocumentQuestionAnsweringPipelineTests::test_large_model_pt_chunk - AssertionError: Lists differ: [{'score': 0.9974, 'answer': '1110212019', 'start': 23, [69 chars] 16}] != [{'score': 0.9967, 'answer': '1102/2019', 'start': 22, '[67 chars] 15}]
FAILED tests/pipelines/test_pipelines_document_question_answering.py::DocumentQuestionAnsweringPipelineTests::test_large_model_pt_donut - TypeError: forward() got an unexpected keyword argument 'page'
FAILED tests/pipelines/test_pipelines_document_question_answering.py::DocumentQuestionAnsweringPipelineTests::test_large_model_pt_layoutlm_chunk - IndexError: index out of range in self
FAILED tests/pipelines/test_pipelines_document_question_answering.py::DocumentQuestionAnsweringPipelineTests::test_pt_LayoutLMConfig_LayoutLMForQuestionAnswering_LayoutLMTokenizerFast_nofeature_extractor - IndexError: tuple index out of range
FAILED tests/pipelines/test_pipelines_document_question_answering.py::DocumentQuestionAnsweringPipelineTests::test_pt_LayoutLMConfig_LayoutLMForQuestionAnswering_LayoutLMTokenizer_nofeature_extractor - IndexError: tuple index out of range
```
  - (For `test_large_model_pt_layoutlm_chunk`, I got different error from our CI where it is just output not equal expected value)
  - The 2 tests with `_chunk` at the end are new tests. The other 3 tests worked well in previous commits before this PR

Could you take a look, please 🙏 ?

Locally, I got one more test failure (not appear on CI), which also fails with previous commit (locally, not on CI). However, you can focus on the above 5 tests.
```
FAILED n tests/pipelines/test_pipelines_document_question_answering.py::DocumentQuestionAnsweringPipelineTests::test_large_model_pt_layoutlm - IndexError: index out of range in self
```

**Comment by ankrgyl** [comment]
I'm so sorry about that! I will take a look ASAP.

**Pull request merged**